src/Microsoft.ML.Data/DataView/ArrayDataViewBuilder.cs

```diff
@@ -85,6 +90,25 @@ public void AddColumn(string name, ValueGetter<VBuffer<ReadOnlyMemory<char>>> ge
             _names.Add(name);
         }
 
+        /// <summary>
+        /// Constructs a new key column from an array where values are copied to output simply
+        /// by being assigned.
+        /// </summary>
+        /// <param name="name">The name of the column.</param>
+        /// <param name="getKeyValues">The delegate that does a reverse lookup based upon the given key. This is for metadata creation</param>
+        /// <param name="keyMin">The <see cref="KeyType"/> minimum to use.</param>
+        /// <param name="keyCount">The count of unique keys specified in values</param>
+        /// <param name="values">The values to add to the column. Note that since this is creating a <see cref="KeyType"/> column, the values will be offset by 1.</param>
+        public void AddColumn(string name, ValueGetter<VBuffer<ReadOnlyMemory<char>>> getKeyValues, ulong keyMin, int keyCount, params ulong[] values)
+        {
+            _host.CheckValue(getKeyValues, nameof(getKeyValues));
+            _host.CheckParam(keyCount > 0, nameof(keyCount));
+            CheckLength(name, values);
+            _columns.Add(new AssignmentColumn<ulong>(new KeyType(DataKind.U8, keyMin, keyCount), values));
```

Comment:
>DataKind.U8 [](start = 65, length = 11)

can this merge with the previous, and the method uses two generics, one for uint/ulong the other for U4/U8 or it would be mistakenly used for other things too?

Reply:
I don't believe that will work, we can use an enum as a generic argument, like DataKind, but DataKind.U8 is a specific value of the DataKind enum.
But also if this did work, it would require the user to specify the DataKind and that they specify the correct data kind when they make the call. What I like about the current implementation is that they pass an array of uint or ulong and correct code will be called, keeping the DataKind internal.

---
In reply to: [243492552](https://github.com/dotnet/machinelearning/pull/1710#discussion_r243492552) [](ancestors = 243492552)

Reply:
actually what you mentioned above would work, I did a slight variation on it though and just get the ItemType from the values. So overall I was able to merge these two. Thank you for bringing this up.

---
In reply to: [243508591](https://github.com/dotnet/machinelearning/pull/1710#discussion_r243508591) [](ancestors = 243508591,243492552)